libraries/AP_Airspeed/AP_Airspeed_SDP3X.cpp

```diff
@@ -0,0 +1,303 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+  backend driver for airspeed from a I2C SDP3X sensor
+
+  with thanks to https://github.com/PX4/Firmware/blob/master/src/drivers/sdp3x_airspeed
+ */
+#include "AP_Airspeed_SDP3X.h"
+
+#include <stdio.h>
+
+#define SDP3X_SCALE_TEMPERATURE		200.0f
+
+#define SDP3XD0_I2C_ADDR            	0x21
+#define SDP3XD1_I2C_ADDR            	0x22
+#define SDP3XD2_I2C_ADDR            	0x23
+
+#define SDP3X_CONT_MEAS_AVG_MODE    	0x3615
+#define SDP3X_CONT_MEAS_STOP        	0x3FF9
+
+#define SDP3X_SCALE_PRESSURE_SDP31	60
+#define SDP3X_SCALE_PRESSURE_SDP32	240
+#define SDP3X_SCALE_PRESSURE_SDP33	20
+
+#define CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C             1.225f                  /* kg/m^3               */
+#define CONSTANTS_AIR_GAS_CONST                         287.1f                  /* J/(kg * K)           */
+#define CONSTANTS_ABSOLUTE_NULL_CELSIUS                 -273.15f                /* °C                   */
+
+extern const AP_HAL::HAL &hal;
+
+AP_Airspeed_SDP3X::AP_Airspeed_SDP3X(AP_Airspeed &_frontend, uint8_t _instance) :
+    AP_Airspeed_Backend(_frontend, _instance)
+{
+}
+
+/*
+  send a 16 bit command code
+ */
+bool AP_Airspeed_SDP3X::_send_command(uint16_t cmd)
+{
+    uint8_t b[2] {uint8_t(cmd >> 8), uint8_t(cmd & 0xFF)};
+    return _dev->transfer(b, 2, nullptr, 0);
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_SDP3X::init()
+{
+    const uint8_t addresses[3] = { SDP3XD0_I2C_ADDR,
+                                   SDP3XD1_I2C_ADDR,
+                                   SDP3XD2_I2C_ADDR
+                                 };
+    bool found = false;
+    bool ret = false;
+
+    for (uint8_t i=0; i<ARRAY_SIZE(addresses) && !found; i++) {
+        _dev = hal.i2c_mgr->get_device(get_bus(), addresses[i]);
+        if (!_dev) {
+            continue;
+        }
+        if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+            continue;
+        }
+
+        // lots of retries during probe
+        _dev->set_retries(10);
+
+        // stop any precending measurements
+        if (!_send_command(SDP3X_CONT_MEAS_STOP)) {
+            _dev->get_semaphore()->give();
+            continue;
+        }
+
+        // start continuous average mode
+        if (!_send_command(SDP3X_CONT_MEAS_AVG_MODE)) {
+            _dev->get_semaphore()->give();
+            continue;
+        }
+
+        hal.scheduler->delay_microseconds(20000);
```

Comment:
Do we really need a delay here? If we really need one for init reasons could we address it with a callback instead? Adding a 20ms pause per airspeed sensor to the boot time seems quite painful.

Reply:
yes, both delays are needed for it to be reliable. I don't think a 40ms delay is really bad though.